File: 신기하/다단계 칫솔 판매.py
'''
테스트 1 〉	통과 (0.03ms, 10.3MB)
테스트 2 〉	통과 (0.12ms, 10.3MB)
테스트 3 〉	통과 (0.08ms, 10.3MB)
테스트 4 〉	통과 (0.20ms, 10.2MB)
테스트 5 〉	통과 (2.21ms, 10.3MB)
테스트 6 〉	통과 (8.95ms, 13.3MB)
테스트 7 〉	통과 (9.32ms, 13.3MB)
테스트 8 〉	통과 (9.77ms, 13.6MB)
테스트 9 〉	통과 (23.63ms, 14.3MB)
테스트 10 〉	통과 (174.77ms, 21.6MB)
테스트 11 〉	통과 (138.53ms, 21.1MB)
테스트 12 〉	통과 (128.92ms, 21.1MB)
테스트 13 〉	통과 (132.95ms, 21.1MB)
'''

def solution(enroll, referral, seller, amount):
    answer = []
    # [자식1, 자식2, 부모, 돈]에 대한 트리 생성
    # 1번 인덱스를 center로 사용하기 때문에 +2만큼 트리 생성
    tree = [[0, 0, 0, 0] for _ in range(len(enroll) + 2)]
    # 조직 구성원에 대해 index를 value로 설정하기 위한 딕셔너리
    # 인덱스를 트리와 맞추기 위해 +2
    enroll_dict = {}
    for i in range(2, len(enroll) + 2):
        enroll_dict[enroll[i - 2]] = i
    
    # 부모, 자식 트리에 적용하기
    for i in range(1, len(referral) + 1):
        # 부모 인덱스 찾기
        if referral[i - 1] == '-': # 부모가 center일 때
            parent = 1
        else: # 부모가 enroll일 때
            parent = enroll_dict[referral[i - 1]]
            
        child = enroll_dict[enroll[i - 1]] # 자식 인덱스 찾기
        
        # 부모 노드에 자식 노드 연결
        tree[child][2] = parent
        if tree[parent][0] == 0: # 자식이 없다면 0번 인덱스에 자식 넣기
            tree[parent][0] = child
        else: # 자식이 있다면 1번 인덱스에 자식 넣기
            tree[parent][1] = child
        

    for i in range(len(seller)):
        money = amount[i] * 100 # 수입
        stack = [enroll_dict[seller[i]]] # 판매원의 트리 인덱스를 찾기 위한 stack 설정
        
        while stack:
            tree_num = stack.pop() # 판매원의 트리 인덱스
            my_money = money - (money // 10) # 본인에게 남는 돈
            money //= 10 # 10% 수수료
            tree[tree_num][3] += my_money # 본인의 이익 넣기
            
            # 백트레킹
            # 올라갈 돈이 없다면 순회하는 의미가 없기 때문에 가지치기
            if money < 1:
                break
            
            # 부모 노드가 있다면 부모로 올라가기 위한 stack
            if tree[tree_num][2] != 0:
                stack.append(tree[tree_num][2])

    for i in range(2, len(tree)):
        answer.append(tree[i][3])
    
    return answer

File: 박민상/210908/다단계칫솔판매.java
import java.util.*;

class Solution {
    // enroll에 있는 이름의 순서를 저장하기위한 map입니다.
    static Map<String, Integer> idxMap;
    // referral을 static으로 사용하기 위한 변수입니다.
    static String[] sreferral;
    static int[] answer;

    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        sreferral = referral;

        idxMap = new HashMap<>();

        //enroll에 있던 이름에 순서를 매겨 map에 저장합니다
        for(int i=0; i<enroll.length; i++){
            idxMap.put(enroll[i], i);
        }

        answer = new int[enroll.length];

        //func함수는 판매자, 판매금액을 인자로 받습니다
        for(int i=0; i<seller.length; i++){
            func(seller[i], amount[i]*100);
        }

        return answer;
    }

    //func는 재귀함수 구조로 본인의 몫을 뗀 후 추천인을 다시 func함수로 호출합니다.
    static void func(String name, int money){
        //기저 조건은 추천인이 민호인 경우입니다.
        if(name.equals("-")) return;

        //현재 판매한 사람의 순번과 부모의 이름을 저장하는 부분입니다.
        int idx=idxMap.get(name);
        String parent = sreferral[idx];

        // int parentmoney=money/10+(money%10>=5?1:0); 처음 코드 작성시 틀렸던 부분입니다. 추천인에게 주는 돈은 절삭하는 것이므로 10으로 나눈 몫입니다.
        int parentmoney=money/10; //추천인에게 줄 돈
        int mymoney=money-parentmoney; //내가 가질 돈은 추천인에게 줄 돈을 뺀 나머지

        answer[idx]+=mymoney; //나는 idx번째이므로 answer[idx]에 내 돈을 추가합니다.

        // 추천인에게 줄 돈이 있는 경우 재귀호출합니다
        if(parentmoney>0){
            func(parent, parentmoney);
        }
    }
}